documentapi/src/main/java/com/yahoo/documentapi/local/LocalVisitorSession.java

```diff
@@ -0,0 +1,164 @@
+        this.outstanding.putAll(access.documents);
+        this.state = new AtomicReference<>(State.RUNNING);
+
+        start();
+    }
+
+    void start() {
+        new Thread(() -> {
+            try {
+                // Iterate through all documents and pass on to data handler
+                outstanding.forEach((id, document) -> {
+                    if (selector.accepts(new DocumentPut(document)) != Result.TRUE)
+                        return;
+
+                    Document copy = new Document(document.getDataType(), document.getId());
+                    for (Field field : document.getDataType().getFields())
+                        if (fieldSet.contains(field))
+                            copy.setFieldValue(field, document.getFieldValue(field));
+
+                    data.onMessage(new PutDocumentMessage(new DocumentPut(copy)),
+                                   new AckToken(id));
+                });
+                // Transition to a terminal state when done
+                state.updateAndGet(current -> {
+                    switch (current) {
+                        case RUNNING:
+                            control.onDone(VisitorControlHandler.CompletionCode.SUCCESS, "Success");
+                            return State.SUCCESS;
+                        case ABORTED:
+                            control.onDone(VisitorControlHandler.CompletionCode.ABORTED, "Aborted by user");
+                            return State.ABORTED;
+                        default:
+                            control.onDone(VisitorControlHandler.CompletionCode.FAILURE, "Unexpected state '" + current + "'");;
+                            return State.FAILURE;
+                    }
+                });
+            }
+            // Transition to failure terminal state on error
+            catch (Exception e) {
+                state.set(State.FAILURE);
+                outstanding.clear();
+                control.onDone(VisitorControlHandler.CompletionCode.FAILURE, Exceptions.toMessageString(e));
+            }
+            finally {
+                data.onDone();
+            }
+        }).start();
+    }
+
+    @Override
+    public boolean isDone() {
+        return    outstanding.isEmpty() // All documents ack'ed
+               && control.isDone();     // Control handler has been notified
+    }
+
+    @Override
+    public ProgressToken getProgress() {
+        throw new UnsupportedOperationException("Progress tokens are not supported");
+    }
+
+    @Override
+    public Trace getTrace() {
+        throw new UnsupportedOperationException("Traces are not supported");
+    }
+
+    @Override
+    public boolean waitUntilDone(long timeoutMs) throws InterruptedException {
+        return control.waitUntilDone(timeoutMs);
+    }
+
+    @Override
+    public void ack(AckToken token) {
+        outstanding.remove((DocumentId) token.ackObject);
```

Comment:
Since `outstanding` is a `ConcurrentSkipListMap` I'm presuming it operates on a logical snapshot of the set of initial document IDs where concurrent removes can't mess anything up? It's been a while since I've looked into its semantics 🙂

Reply:
Its iterators are [weakly consistent](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#Weakly) — while this is weakly defined, they _can_ reflect changes during iteration.
This is the case for the last part of the unit test. I suppose this test would break, were the implementation of `ConcurrentSkipListMap` ever to change.

Reply:
Put a guard on state being `RUNNING` in the iteration.